---
title: "MISKmeans"
author: "Caleb"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MISKmeans}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


## Introduction

MISKmeans is short for meta-analytic integrative sparse Kmeans, which aims to integrate multiple omics data, multiple cohort data and incorporate prior biological knowledge to perform feature selection and sample clustering (discovery disease subtypes).
The input should be multiple cohorts, with each cohort containing one or more omics profiles (e.g. gene expression, CNV, methylation...).
Usually the prior biological knowledge could be the cis regulatory relationship between mRNA, methylation and CNV of the same gene symbol;
or a biological pathway including a collection of functional genes which can potentially overlap with other pathways.
In the next section, we will provide a concrete example how to use MISKmeans package.

## Usage of MISKmeans package

Below, we simulate 2 studies with each study contains 3 subtypes and 1000 genes.
Gene 1 to gene 100 are intrinsic genes.
The group structure is defined as following:
1 - 51, 2 - 52, ..., 50 - 100, totally 50 groups.


```{r}
library(MISKmeans, quietly=FALSE)

S <- 2 ## number of studies
K <- 3 ## number of subtypes
G <- 1000 ## total number of genes
g1 <- 50 ## number of intrinsic gene set 1
g2 <- 50 ## number of intrinsic gene set 2
n0 <- 20 ## nunmber of subjects per subtype
n <- K*n0 ## total nunmber of subjects per study
labels <- cut(1:n,breaks=K,labels=FALSE) ## generate clustering membership

set.seed(32611)
S1 <- matrix(rnorm(G*n), nrow=G, ncol=n)
S2 <- matrix(rnorm(G*n), nrow=G, ncol=n)

S1[1:g1, labels==1] <- S1[1:g1, labels==1] + 2
S1[1:g1, labels==3] <- S1[1:g1, labels==3] - 2
S1[g1 + 1:g2, labels==1] <- S1[g1 + 1:g2, labels==1] - 2
S1[g1 + 1:g2, labels==2] <- S1[g1 + 1:g2, labels==2] + 2

heatmap(S1, Rowv=NA, Colv=NA) ## visualize study 1

S2[1:g1, labels==2] <- S2[1:g1, labels==2] + 2
S2[1:g1, labels==1] <- S2[1:g1, labels==1] - 2
S2[g1 + 1:g2, labels==2] <- S2[g1 + 1:g2, labels==2] - 2
S2[g1 + 1:g2, labels==3] <- S2[g1 + 1:g2, labels==3] + 2

heatmap(S2, Rowv=NA, Colv=NA) ## visualize study 1

S = list(t(S1),t(S2)) ## prepare input
groups <- Map('c',1:g1,g1 + 1:g2) ## prepare prior group structure

res <- MISKmeans(d = S, K = 3, gamma = 0.4, group = groups) ## perform MISKmeans. gamma is the tuning parameter to control the level of penalty.

names(res)

order_S1 <- order(res$Cs[[1]])
order_S2 <- order(res$Cs[[2]])

S1_sub <- S1[res$ws!=0, order_S1]
S2_sub <- S2[res$ws!=0, order_S2]

col_S1 <- palette()[res$Cs[[1]][order_S1]]
col_S2 <- palette()[res$Cs[[2]][order_S2]]

heatmap(S1_sub, Rowv=NA, Colv=NA, ColSideColors=col_S1) ## visualize the subtyping result for S1
heatmap(S2_sub, Rowv=NA, Colv=NA, ColSideColors=col_S2) ## visualize the subtyping result for S2
```


